Plugins/NativeEngine/Source/NativeEngine.cpp

```diff
@@ -604,12 +638,77 @@ namespace Babylon
 
         const bgfx::Attrib::Enum attrib = static_cast<bgfx::Attrib::Enum>(location);
         const bgfx::AttribType::Enum attribType = static_cast<bgfx::AttribType::Enum>(type);
-        vertexLayout.add(attrib, static_cast<uint8_t>(numElements), attribType, normalized);
-        vertexLayout.m_stride = static_cast<uint16_t>(byteStride);
+
+        const bool promoteToFloats = (bgfx::getCaps()->rendererType == bgfx::RendererType::Direct3D11 || 
+            bgfx::getCaps()->rendererType == bgfx::RendererType::Direct3D12 || 
+            bgfx::getCaps()->rendererType == bgfx::RendererType::Vulkan) && 
+            (attribType == bgfx::AttribType::Int8 || 
+            attribType == bgfx::AttribType::Uint8 || 
+            attribType == bgfx::AttribType::Uint10 || 
+            attribType == bgfx::AttribType::Int16 || 
+            attribType == bgfx::AttribType::Uint16) &&
+            (attrib == bgfx::Attrib::Position ||
+            attrib == bgfx::Attrib::Normal ||
+            attrib == bgfx::Attrib::Bitangent ||
+            attrib == bgfx::Attrib::Tangent);
+
+        if (promoteToFloats)
+        {
+            vertexLayout.add(attrib, static_cast<uint8_t>(numElements), bgfx::AttribType::Float, normalized);
+            vertexLayout.m_stride = static_cast<uint16_t>(sizeof(float) * numElements);
+            if (normalized)
+            {
+                switch (attribType)
+                {
+                case bgfx::AttribType::Int8:
+                    vertexBufferData->PromoteToFloats<int8_t, true>(numElements, byteStride, 127.f);
+                    break;
+                case bgfx::AttribType::Uint8:
+                    vertexBufferData->PromoteToFloats<uint8_t, true>(numElements, byteStride, 255.f);
+                    break;
+                case bgfx::AttribType::Int16:
+                    vertexBufferData->PromoteToFloats<int16_t, true>(numElements, byteStride, 32767.f);
+                    break;
+                case bgfx::AttribType::Uint16:
+                    vertexBufferData->PromoteToFloats<uint16_t, true>(numElements, byteStride, 65535.f);
+                    break;
+                case bgfx::AttribType::Uint10: // is supported by any format ?
+                default:
+                    throw std::runtime_error("Unable to promote vertex stream to a float array.");
+                }
```

Comment:
I don't think you need this switch. You can use templates with [std::numeric_limits<T>::max](https://en.cppreference.com/w/cpp/types/numeric_limits/max) to achieve the same effect and pass normalized as a regular parameter.

Reply:
Added numeric_limits. The normalized boolean template is useful for compile time check instead of runtime branching.